### PR TITLE
chore: Add new UserSettings store [WEB-1353]

### PR DESCRIPTION
**Pull request by ClaireNeveu**
## Description
This PR adds a new store that provides a more idiomatic way of reading and writing user settings. It temporarily shares state with the old `useSettings` hook until that is replaced with this. Main features:
- Reduced complexity (no checking equality before writing, handling query params, etc)
- Improved serialization (settings can be anything)
- Impossible to accidentally access settings while they're still loading

## Test Plan
- Check if all settings data is fetched when page is loaded
- Check if there no infinite settings calls in network tab
- Check if settings POST call is invoked when user changes the settings and check DB
  - ex: changing filters in the new exp table, close filter popup, refresh page, check if filters persist
- Check if there's no console error in console tab (should be debug mode)

## Checklist

- [ ] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket
WEB-1353

### CODE REVIEW

**Review by keita-determined — COMMENTED** [review]
List the issue(s) that would be caused by this PR. Ignored other usesettings related bugs that has been there before this PR

Overall i didnt find many bugs.

1. Saw this `Warning: Maximum update depth exceeded. This can happen when a component calls setState inside useEffect, but useEffect either doesn't have a dependency array, or one of the dependencies changes on every render.` which i didnt see in main branch in cluster page.


https://github.com/determined-ai/determined/assets/109113616/4764fd33-b090-484f-a726-5234d319f968

---

**Comment by johnkim-det** [comment]
tested pages where `useSettings` is used and looks like settings are getting read/written as expected. FWIW not reproducing the resource pool page issue on netlify preview.

**Comment by keita-determined** [comment]
> tested pages where `useSettings` is used and looks like settings are getting read/written as expected. FWIW not reproducing the resource pool page issue on netlify preview.

You wouldnt see the console warning cuz its production mode on Netflify.

**Comment by ClaireNeveu** [comment]
@keita-determined This one was a little mysterious. Root cause here is [JobQueue creating a new settings config every render](https://github.com/determined-ai/determined/pull/7322/files#diff-0c9970101540a65454abd88ef54eee041f625029c1e8079bdcf5b9bb2c521465R71) but I would expect that to have been an issue with the old code as well. I also fixed [another line of code](https://github.com/determined-ai/determined/pull/7322/commits/8be61b907aa3307016463f1a5c0e1c39c334d7e0#diff-0ac40df189229b81aa5948c8d37e1c96bea9e253bdf2b233fef43661829242f3R167) that I wrote which should have caused an issue but wasn't.

@johnkim-det Can you review for code quality and comprehensibility?

File: webui/react/src/hooks/useSettings.test.tsx [hunk 1]
```diff
@@ -17,6 +22,69 @@ vi.mock('services/api', () => ({
...
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isTypeC(codec: t.Encoder<any, any>): codec is t.TypeC<t.Props> {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  return (codec as any)._tag === 'InterfaceType';
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type Json = any;
+type State = Map<string, Record<string, unknown>>;
```
Comment (johnkim-det) [inline webui/react/src/hooks/useSettings.test.tsx]: import types from store?

**Review by johnkim-det — APPROVED** [review]

File: webui/react/src/stores/userSettings.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+    this.updateUserSetting(key, null);
+    this.#settings.update((loadable) => {
+      return Loadable.map(loadable, (map) => {
+        return map.removeAll(key);
+      });
+    });
+  }
+
+  public reset(): void {
+    this.#settings.set(NotLoaded);
```
[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/stores/userSettings.tsx:92]: This is for when the user logs out.

[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/stores/userSettings.tsx:108]: Actually added it to the signout in a new commit

File: webui/react/src/pages/JobQueue/JobQueue.tsx [hunk 3]
```diff
@@ -68,7 +68,9 @@ const JobQueue: React.FC<Props> = ({ selectedRp, jobState }) => {
   const [pageState, setPageState] = useState<{ isLoading: boolean }>({ isLoading: true });
   const pageRef = useRef<HTMLElement>(null);
 
-  const { settings, updateSettings } = useSettings<Settings>(settingsConfig(jobState));
+  const { settings, updateSettings } = useSettings<Settings>(
```
Comment (ClaireNeveu) [inline webui/react/src/pages/JobQueue/JobQueue.tsx:71]: Incidental bug that was uncovered due to this PR. The config passed to `useSettings` needs to be a stable reference.

**Pull request merged**